assembly/index.ts

```diff
@@ -11,100 +9,95 @@ export function pixelmatch(
 	height: i32,
 
 	// Can't use interfaces to make an options object here: https://docs.assemblyscript.org/basics/limitations#oop
-	threshold: f32,     	// matching threshold (0 to 1); smaller is more sensitive
-	includeAA: bool,    	// whether to skip anti-aliasing detection
-	alpha: f32,				// opacity of original image in diff ouput
-	aaR: f32, 				// r color of anti-aliased pixels in diff output
-	aaG: f32,				// g color of anti-aliased pixels in diff output
-	aaB: f32,				// b color of anti-aliased pixels in diff output
-	diffR: f32,				// r color of different pixels in diff output
-	diffG: f32,				// g color of different pixels in diff output
-	diffB: f32				// b color of different pixels in diff output
+	threshold: f64,   // matching threshold (0 to 1); smaller is more sensitive
+	includeAA: bool,  // whether to skip anti-aliasing detection
+	alpha: f64,       // opacity of original image in diff ouput
+	aaR: f64,         // r color of anti-aliased pixels in diff output
+	aaG: f64,         // g color of anti-aliased pixels in diff output
+	aaB: f64,         // b color of anti-aliased pixels in diff output
+	diffR: f64,       // r color of different pixels in diff output
+	diffG: f64,       // g color of different pixels in diff output
+	diffB: f64        // b color of different pixels in diff output
 ): i32 {
 
 	// No Errors: https://docs.assemblyscript.org/basics/limitations#exceptions
 
 	// Image sizes do not match
 	if (img1.length !== img2.length || (output !== null && output.length !== img1.length)) {
-		return -1; 
+		return -1;
 	}
 
 	// Image data size does not match width/height.
 	if (img1.length !== width * height * 4) {
 		return -2;
 	}
 
+  let img1Ptr = img1.dataStart;
+  let img2Ptr = img2.dataStart;
+  let outputPtr = output.dataStart;
+
 	threshold = isNaN(threshold) ? 0.1 : threshold;
 	includeAA = includeAA || false;
-	alpha = isNaN(alpha) ? 0.1 : alpha;
+	alpha     = isNaN(alpha) ? 0.1 : alpha;
 
 	// check if images are identical
-	const len: i32 = width * height;
-	let identical: bool = true;
-
-	for (let i = 0; i < len; i++) {
-		if (img1[i] !== img2[i]) {
-			identical = false;
-			break;
-		}
-	}
+	let len = width * height * 4;
+  let identical = memory.compare(img1Ptr, img2Ptr, len) === 0;
 
 	// fast path if identical
-	if (identical) { 
+	if (identical) {
 		if (output) {
-			for (let i = 0; i < len; i++) {
-				drawGrayPixel(img1, 4 * i, alpha, output);
+			for (let i = 0; i < len; i += 4) {
```

Comment:
I'm also curious why it couldn't be:

```ts
for (let ptr = img1Ptr, end = ptr + len; ptr < end; ptr += 4) {
  drawGrayPixel(ptr, alpha, outputPtr);
}
```

Each iteration can now load one less value.

This does mean refactoring the way `drawGrayPixel()` works too.

EDIT: A while loop would be fine here too.

Reply:
I don't think this refactoring change something because `drawGrayPixel` already inlined.

Reply:
I was saying each loop iteration needed to load 4 * n values instead of 3 * n + 1.

Reply:
I understand now :). Thanks for the info.